selfdrive/test/process_replay/process_replay.py

```diff
@@ -206,204 +191,291 @@ def get_car_params(msgs, fsm, can_sock, fingerprint):
     timeout=60*10,  # first messages are blocked on internet assistance
+    drained_pub="ubloxGnss", # config_callback will switch this to qcom if needed 
   ),
   ProcessConfig(
     proc_name="torqued",
-    pub_sub={
-      "liveLocationKalman": ["liveTorqueParameters"],
-      "carState": [], "controlsState": [],
-    },
+    pubs=["liveLocationKalman", "carState", "carControl"],
+    subs=["liveTorqueParameters"],
     ignore=["logMonoTime"],
-    init_callback=get_car_params,
+    config_callback=None,
+    init_callback=get_car_params_callback,
     should_recv_callback=torqued_rcv_callback,
     tolerance=NUMPY_TOLERANCE,
-    fake_pubsubmaster=True,
+    polled_pubs=["liveLocationKalman"],
   ),
 ]
 
 
 def replay_process(cfg, lr, fingerprint=None):
   with OpenpilotPrefix():
-    if cfg.fake_pubsubmaster:
-      return python_replay_process(cfg, lr, fingerprint)
+    controlsState = None
+    initialized = False
+    for msg in lr:
+      if msg.which() == 'controlsState':
+        controlsState = msg.controlsState
+        if initialized:
+          break
+      elif msg.which() == 'carEvents':
+        initialized = car.CarEvent.EventName.controlsInitializing not in [e.name for e in msg.carEvents]
+
+    assert controlsState is not None and initialized, "controlsState never initialized"
+
+    CP = [m for m in lr if m.which() == 'carParams'][0].carParams
+    if fingerprint is not None:
+      setup_env(cfg=cfg, controlsState=controlsState, lr=lr, fingerprint=fingerprint)
     else:
-      return replay_process_with_sockets(cfg, lr, fingerprint)
+      setup_env(CP=CP, cfg=cfg, controlsState=controlsState, lr=lr)
+
+    if cfg.config_callback is not None:
+      params = Params()
+      cfg.pubs, cfg.drained_pub = cfg.config_callback(params, cfg)
+      cfg.polled_pubs = list(set(cfg.polled_pubs) & set(cfg.pubs))
+
+    all_msgs = sorted(lr, key=lambda msg: msg.logMonoTime)
+    pub_msgs = [msg for msg in all_msgs if msg.which() in set(cfg.pubs)]
+
+    with ReplayContext(cfg) as rc:
+      pm = messaging.PubMaster(cfg.pubs)
+      sockets = {s: messaging.sub_sock(s, timeout=100) for s in cfg.subs}
 
+      managed_processes[cfg.proc_name].prepare()
+      managed_processes[cfg.proc_name].start()
 
-def setup_env(simulation=False, CP=None, cfg=None, controlsState=None, lr=None):
+      if cfg.init_callback is not None:
+        cfg.init_callback(rc, pm, all_msgs, fingerprint)
+        CP = car.CarParams.from_bytes(Params().get("CarParams", block=True))
+
+      log_msgs, msg_queue = [], []
+      try:
+        # Wait for process to startup
+        with Timeout(10, error_msg=f"timed out waiting for process to start: {repr(cfg.proc_name)}"):
+          while not all(pm.all_readers_updated(s) for s in cfg.pubs):
+            time.sleep(0)
+
+        for s in sockets.values():
+          messaging.recv_one_or_none(s)
+
+        # TODO wait for sockets to reconnect, for now lets just sleep
+        time.sleep(1)
```

Comment:
can't we directly check this?

Reply:
I've found a way to handle that differently, by calling receive on sub-sockets for the first time, when I'm sure the process is ready (recv_called set for the first time)